csharp/src/Google.Protobuf.Benchmarks/SerializationBenchmark.cs

```diff
@@ -0,0 +1,119 @@
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#endregion
+
+using BenchmarkDotNet.Attributes;
+using System.IO;
+using System.Linq;
+
+namespace Google.Protobuf.Benchmarks
+{
+    /// <summary>
+    /// Benchmark for serializing (to a MemoryStream) and deserializing (from a ByteString).
+    /// Over time we may wish to test the various different approaches to serialization and deserialization separately.
+    /// </summary>
+    public class SerializationBenchmark
+    {
+        /// <summary>
+        /// All the configurations to be tested. Add more datasets to the array as they're available.
+        /// (When C# supports proto2, this will increase significantly.)
+        /// </summary>
+        public static SerializationConfig[] Configurations => new[]
+        {
+            new SerializationConfig("dataset.google_message1_proto3.pb")
+        };
+
+        [ParamsSource(nameof(Configurations))]
+        public SerializationConfig Configuration { get; set; }
+
+        private MessageParser parser;
+        /// <summary>
+        /// Each data set can contain multiple messages in a single file.
+        /// Each "write" operation should write each message in turn, and each "parse"
+        /// operation should parse each message in turn.
+        /// </summary>
+        private SubTest[] subTests;
+
+        [GlobalSetup]
+        public void GlobalSetup()
+        {
+            parser = Configuration.Parser;
+            subTests = Configuration.Payloads.Select(p => new SubTest(p, parser.ParseFrom(p))).ToArray();
+        }
+
+        [IterationSetup]
+        public void IterationSetup()
+        {
+            foreach (var item in subTests)
+            {
+                item.Reset();
+            }
+        }
+
+        [Benchmark]
+        public void WriteToStream()
+        {
+            foreach (var item in subTests)
```

Comment:
Does this aggregate all the tested scenarios into one result? It would be useful to see grauar results, e.g. serializing a small message was fast but a message with a large large string was slower

Reply:
The idea is that each scenario *can* contain multiple messages. As per benchmarks.proto:

```text
  // The payload(s) for this dataset.  They should be parsed or serialized
  // in sequence, in a loop, ie.
  //
  //  while (!benchmarkDone) {  // Benchmark runner decides when to exit.
  //    for (i = 0; i < benchmark.payload.length; i++) {
  //      parse(benchmark.payload[i])
  //    }
  //  }
  //
  // This is intended to let datasets include a variety of data to provide
  // potentially more realistic results than just parsing the same message
  // over and over.  A single message parsed repeatedly could yield unusually
  // good branch prediction performance.
```

For more granular results, you'd have separate datasets with one payload each. It's fine to have multiple datasets for the same message.